File: ejercicio4.py
def main() -> None:
    # Matrices y arreglos
    stock: list[list[int]] = []
    distancia: list[list[int]] = []
    umbral: list[int] = [10 for _ in range(15)]

    # Contadores
    total_movimientos = 0
    costo_total = 0

    # Llenar stock (datos de prueba)
    for i in range(6):
        nueva: list[int] = []
        for _ in range(15):
            nueva.append(5 + i * 5)
        stock.append(nueva)

    # Llenar distancia
    for i in range(6):
        nueva: list[int] = []
        for j in range(6):
            nueva.append(abs(i - j) + 1)
        distancia.append(nueva)

    for i in range(15):
        for j in range(6):
            if not necesita(stock[j][i], umbral[i]):
                continue

            # Si necesita producto
            # Buscar quien tiene excedente
            for k in range(6):
                if excedente(stock[k][i], umbral[i]) <= 0:
                    continue

                falta = umbral[i] - stock[j][i]
                sobra = stock[k][i] - umbral[i]

                mover = min(falta, sobra)
                costo = mover * distancia[k][j]

                # Actualizar
                stock[j][i] += mover
                stock[k][i] -= mover

                total_movimientos += 1
                costo_total += costo

                # Mostrar producto
                print(f"Producto {i}: {mover} unidades movidas de almacen {k} a almacen {j}. Costo: {costo}")

                # Si ya no necestia más, salir
                if stock[j][i] >= umbral[i]:
                    break

    print("-----------------------")
    print(f"Total de movimientos: {total_movimientos}")
    print(f"Costo total: {costo_total}")
    

def necesita(cantidad: int, umbral: int) -> bool:
    return cantidad < umbral


def excedente(cantidad: int, umbral: int) -> int:
    if cantidad > umbral:
        return cantidad - umbral
    return 0


if __name__ == "__main__":
    main()

File: ejercicio4.cpp
#include <iostream>
#include <cstdlib>   // abs
#include <algorithm> // min
using namespace std;

// Funciones
bool necesita(int cantidad, int umbral);
int excedente(int cantidad, int umbral);

int main() {
    // Matrices y arreglos
    int stock[6][15];
    int distancia[6][6];
    int umbral[15];

    // Contadores
    int totalMovimientos = 0;
    int costoTotal = 0;

    // Llenar umbral
    for (int p = 0; p < 15; p++) {
        umbral[p] = 10;
    }

    // Llenar stock (datos de prueba)
    for (int i = 0; i < 6; i++) {
        for (int p = 0; p < 15; p++) {
            stock[i][p] = 5 + i * 5;
        }
    }

    // Llenar distancias
    for (int i = 0; i < 6; i++) {
        for (int j = 0; j < 6; j++) {
            distancia[i][j] = abs(i - j) + 1;
        }
    }

    cout << "=== INICIO DEL PROGRAMA ===" << endl;

    // ALGORITMO PRINCIPAL
    for (int p = 0; p < 15; p++) {
        for (int i = 0; i < 6; i++) {

            // Si necesita producto
            if (necesita(stock[i][p], umbral[p])) {

                // Buscar quien tiene excedente
                for (int j = 0; j < 6; j++) {

                    if (excedente(stock[j][p], umbral[p]) > 0) {

                        int falta = umbral[p] - stock[i][p];
                        int sobra = stock[j][p] - umbral[p];

                        int mover = min(falta, sobra);
                        int costo = mover * distancia[j][i];

                        // Actualizar
                        stock[i][p] += mover;
                        stock[j][p] -= mover;

                        totalMovimientos++;
                        costoTotal += costo;

                        // Mostrar movimiento
                        cout << "Producto " << p
                             << ": " << mover
                             << " unidades movidas de Almacen " << j
                             << " a Almacen " << i
                             << ". Costo: " << costo << endl;

                        // Si ya no necesita más, salir
                        if (stock[i][p] >= umbral[p]) {
                            break;
                        }
                    }
                }
            }
        }
    }

    // Resultado final
    cout << "---------------------------" << endl;
    cout << "Total de movimientos: " << totalMovimientos << endl;
    cout << "Costo total: " << costoTotal << endl;

    return 0;
}

// Función necesita
bool necesita(int cantidad, int umbral) {
    return cantidad < umbral;
}

// Función excedente
int excedente(int cantidad, int umbral) {
    if (cantidad > umbral)
        return cantidad - umbral;
    return 0;
}